### PR TITLE
Fix c-style main functions

### LINKED ISSUE #12569: exporting a C main function regressed
### Zig Version

0.10.0-dev.3665+9cf667a21

### Steps to Reproduce

```zig
const string = [*:0]const u8;
const stringarray = [*:null]const ?string;

pub export fn main(argc: c_int, argv: stringarray) c_int {
    _ = argc;
    _ = argv;
    return 0;
}
```

### Expected Behavior

successful compile and run

### Actual Behavior

```
/home/meghan/src/zig/lib/std/start.zig:571:17: error: expected return type of main to be 'void', '!void', 'noreturn', 'u8', or '!u8'
                @compileError(bad_main_ret);
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/meghan/src/zig/lib/std/start.zig:510:12: note: called from here
    return @call(.{ .modifier = .always_inline }, callMain, .{});
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/meghan/src/zig/lib/std/start.zig:460:36: note: called from here
    return initEventLoopAndCallMain();
           ~~~~~~~~~~~~~~~~~~~~~~~~^~
/home/meghan/src/zig/lib/std/start.zig:424:17: note: called from here
    std.os.exit(@call(.{ .modifier = .always_inline }, callMainWithArgs, .{ argc, argv, envp }));
                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

### PR DESCRIPTION
**Pull request by gooncreeper**
Fixes #12569

Adds support for main functions with signatures
```zig
pub fn main() c_int
pub fn main(argc: c_int, argv: [*:null]?[*:0]c_char) c_int
pub fn main(argc: c_int, argv: [*:null]?[*:0]c_char, envp: [*:null]?[*:0]c_char) c_int
```

### CODE REVIEW

**Comment by gooncreeper** [comment]
I am unsure why the CI is failing

**Comment by alexrp** [comment]
This will need a rebase.

**Review by andrewrk — CHANGES_REQUESTED** [review]
it's not supposed to work without `export`.

File: lib/std/start.zig [hunk 1]
```diff
@@ -589,30 +592,95 @@ fn main(c_argc: c_int, c_argv: [*][*:0]c_char, c_envp: [*:null]?[*:0]c_char) cal
...
+            if (main_fn.params.len == 2)
+                break :blk root.main(c_argc, c_argv.ptr);
+
+            const c_envp = opt_c_envp orelse (cEnvpAlloc(arena) catch |err|
+                std.debug.panic("failed to allocate envp: {s}", .{@errorName(err)})).ptr;
+
+            if (main_fn.params.len == 3)
+                break :blk root.main(c_argc, c_argv.ptr, c_envp);
+
+            @compileError(bad_c_main);
```
Comment (andrewrk) [inline lib/std/start.zig:680]: this looks wrong, none of this was needed before the regression.

**Review by andrewrk — CHANGES_REQUESTED** [review]
please provide a PR description to justify these sus changes

File: lib/std/start.zig (hunk 1 above)
Comment (gooncreeper) [reply lib/std/start.zig:680]: I agree, taking a quick look at the `0.9.x` branch it didn't have this logic so I wounder how it worked before. I will take a look at this tomorrow.

**Comment by gooncreeper** [comment]
I misunderstood the issue, this should only work when linking against libc.

**Pull request closed without merging**